src/ThreeEditor/util/hooks/signals.tsx

```diff
@@ -0,0 +1,110 @@
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { Object3D } from 'three';
+import { Editor } from '../../js/Editor';
+
+type SignalType = keyof Editor['signals'];
+
+export const useSignalObjectChanged = (
+	editor: Editor,
+	callback: (object: Object3D, property: keyof Object3D) => void
+) => {
+	useSignal(editor, 'objectChanged', callback);
+};
+
+export const useSignal = (
+	editor: Editor,
+	signal: SignalType | SignalType[],
+	callback: (object: Object3D, ...args: any[]) => void
+) => {
+	useEffect(() => {
+		let signalArray = Array.isArray(signal) ? signal : [signal];
+		signalArray.forEach(signal => editor.signals[signal].add(callback));
+		return () => {
+			signalArray.forEach(signal => editor.signals[signal].remove(callback));
+		};
+	}, [callback, editor.signals, signal]);
+};
+
+type ProxyState<T> = T & {
+	object: T;
+};
```

Comment:
Type of object wrapped in `Proxy`, that returns original object under `object` property.

Reply:
It might break things if T already has a property `object`. Will we get a warning that we can't use this type or will it silently pass by the type checker and break something down the line?

Reply:
Something like this bothers me:
![image](https://user-images.githubusercontent.com/26445545/199758559-911e32d0-3dae-4897-8d20-6832b7a5554f.png)

Reply:
Something like this 
```
type ValidState<T> = T extends { object: any } ? never : T;
```
will add an error to the previous example
![image](https://user-images.githubusercontent.com/26445545/199769666-214e2f2e-4e45-4d59-9240-ba88c49a5058.png)

You need to wrap watchedObject type in ValidState just like this:
```
function useSmartWatchEditorState<T>(
	editor: Editor,
	watchedObject: ValidState<NonNullable<T>>,
	watchAnyChange?: boolean,
	debug?: boolean
): { state: ProxyState<T> };
function useSmartWatchEditorState<T>(
	editor: Editor,
	watchedObject: ValidState<NonNullable<T>> | null,
	watchAnyChange?: boolean,
	debug?: boolean
): { state: ProxyState<T> | null };
function useSmartWatchEditorState<T>(
	editor: Editor,
	watchedObject: ValidState<T> | null,
	watchAnyChange = false,
	debug = false
): { state: unknown } {
```

Reply:
ok